### ISSUE TITLE
Overloading the >>- operation to do both map and flatMap

### ISSUE THREAD

**Issue by yoavschwartz** [antitypical/Result#110]
I don't think theres a reason why only flatMap should be tied to the >>- operator. I think it's pretty clear that if you use the operator you want to map/flatMap depending on what you are returning from your function. I find it unlikely that someone would consciously want to map to `Result<Result<T, ErrorTpe>,ErrorType>` just like its probably an error if you map to `Optional<Optional<T>>` and you probably wanted to use flatMap. For extreme cases, they can just type the function name.

**Comment by robrix** [comment]
I’m disinclined to overload `>>-` to perform map. In no particular order:

1. Overloading increases the amount of work Swift has to do in order to compile an expression. Specifically, it necessitates the addition of constraints which can cause “Expression was too complex to be solved in reasonable time” errors to crop up. Every additional overload of any term used in a given expression increases this cost significantly.

2. Overloading increases the amount of semantic work a reviewer or maintainer has to do in order to understand an expression.

3. Nested results and optionals are completely valid use cases whether or not you or I encounter them. We shouldn’t penalize someone who wants to do something using nested results.

4. The distinction between `map` and `flatMap` is extremely important as a reviewer. If something `map`s, you know it’s going to be 1:1—it will not e.g. increase the size of the output. If it `flatMap`s, it warrants more thorough review since it has the capacity for choice: it can turn successes into failures or vice versa. This is extremely important, and would not be made easier by conflating `>>-` to do both jobs.

TL;DR: `map` and `flatMap` represent different _classes_ of operation, and thus should not be conflated in a single operator :-1:

**Comment by yoavschwartz** [comment]
Ok :) sounds good to me
